[#19] Exclude config ignored?

klappradla opened the pull request:
Quickly threw together some of the things @pmeinhardt and I came up in #19 
- require a rubocop version starting at `0.57.0` in order to be able to use `inherit_merge` to merge configuration with rubocop's defaults
- update our list of `AllCops:Exclude` (again) to only list things not already in rubocop's default list
- add a post install message to tell the users they should take care of the behavior of `AllCops:Exclude`

The post install message is a bit cheap for now and looks like this:

```
❯ gem install rubocop-bitcrowd-1.3.0.gem
Fetching: rubocop-0.57.2.gem (100%)
Successfully installed rubocop-0.57.2
PLEASE NOTE: as of version 0.57.0 rubocop allows merging its default
AllCops:Include and AllCops:Exclude lists with user defined lists.

Therefore rubocop-bitcrowd is no longer fully overriding AllCops:Exclude,
only adding some extra paths.

If want to keep excluding everthing as before, add this to your .rubocop.yml:
inherit_mode:
  merge:
    - Exclude

For more details: https://rubocop.readthedocs.io/en/latest/configuration/#merging-arrays-using-inherit_mode

Cheers!
Your friends at bitcrowd
Successfully installed rubocop-bitcrowd-1.3.0
2 gems installed
```

☝️ open for suggestions on improving it 😉

klappradla committed:
Update readme with note about merging defaults

Since rubucoop 0.57.0 one does no longer have to fully override the
AllCops:Include and AllCops:Exclude lists, but can instead merge them.

klappradla committed:
Update the version of rubocop we require

Since 0.57.0, it's possible to "merge" the default configuration for
e.g. AllCops:Include - which is what we usually like (otherwise our gem
would need to fully override the whole list...)

klappradla committed:
Update list of excluded files

Since we're now requiring a rubocop version starting fromm 0.57.0, we
can drop the things that are in the default configuration (see here:
https://github.com/rubocop-hq/rubocop/blob/5d5e7c0f8e195315d10b165e11399b6641754ce2/config/default.yml).

Instead we can now make use of "inherit_mode" to merge the configuration
with the defaults.

klappradla committed:
Add a post install message

I added a post install message to remind users we're changing the
configuration for which directories to include/exclude.

closes #19

klappradla committed:
Add blank lines to the post install message

klappradla commented:
Ok, figured out, it's probably better to add a bit of whitespace to the post-install message ⛄️ 

Looks like this now:
```
❯ gem install rubocop-bitcrowd-1.3.0.gem

PLEASE NOTE: as of version 0.57.0 rubocop allows merging its default
AllCops:Include and AllCops:Exclude lists with user defined lists.

Therefore rubocop-bitcrowd is no longer fully overriding AllCops:Exclude,
only adding some extra paths.

If want to keep excluding everthing as before, add this to your .rubocop.yml:
inherit_mode:
  merge:
    - Exclude

For more details: https://rubocop.readthedocs.io/en/latest/configuration/#merging-arrays-using-inherit_mode

Cheers!
Your friends at bitcrowd

Successfully installed rubocop-bitcrowd-1.3.0
1 gem installed
```

pmeinhardt commented on .rubocop.yml:
```diff
@@ -1,7 +1,5 @@
 AllCops:
   Exclude:
-    - 'node_modules/**/*'
-    - 'vendor/bundle/**/*'
```
I think we should annotate our config a bit more… I was wondering why you removed these lines. Now I quickly found out that these are the defaults (because we've been talking about merging here). In the future though, I think it might not be super-obvious which files/patterns will be ignored.

Maybe add a comment:

```yml
# Add additional exclude patterns for files that should not be analyzed.
# These extend the default RuboCop exclude patterns:
# https://github.com/rubocop-hq/rubocop/blob/master/config/default.yml
AllCops:
  Exclude:
    # …
```

pmeinhardt commented on README.md:
```diff
@@ -22,6 +22,12 @@ To use the configuration in your project create a .rubocop.yml with:
+inherit_mode:
+  merge:
+    - Include
+    - Exclude
```
👍

pmeinhardt commented on rubocop-bitcrowd.gemspec:
```diff
@@ -1,22 +1,41 @@
+    Therefore rubocop-bitcrowd is no longer fully overriding AllCops:Exclude,
+    only adding some extra paths.
+
+    If want to keep excluding everthing as before, add this to your .rubocop.yml:
```
Small typo: "ever**y**thing"

pmeinhardt commented on rubocop-bitcrowd.gemspec:
```diff
@@ -1,22 +1,41 @@
+    AllCops:Include and AllCops:Exclude lists with user defined lists.
+
+    Therefore rubocop-bitcrowd is no longer fully overriding AllCops:Exclude,
+    only adding some extra paths.
```
For some reason this section/the "Therefore" here confused me a bit 🤔 Maybe it's just me.

How about something along the lines of…

```
This new version of rubocop-bitcrowd no longer overrides RuboCops default AllCops:Exclude list,
but only contains extra patterns not included in the defaults.

Therefore if you want to keep excluding both the bitcrowd patterns as well as the RuboCop default ones,
add this to your .rubocop.yml
```

?

klappradla replied:
I adapted it somewhat like this.

pmeinhardt replied:
Neat! I like your version even better ❤️

pmeinhardt reviewed (approved):
Thank you @klappradla for this PR ❤️

Now I also learned where these gem install messages are defined 🙌

klappradla committed:
Add comment to explain exclude patterns

klappradla committed:
Update post install message

Make the post install message a bit clearer.

klappradla committed:
Remove redundant configuration

This removes the `DisplayCopNames` rule. It defaults to true anyways, so
there's not need to define it on our side as well.

pmeinhardt commented:
Thank you @klappradla for these great tweaks. I'd say: :shipit:

tessi reviewed (approved):
looks great. a CHANGELOG.md entry would be appreciated though :)

klappradla committed:
Add a section to the Changelog

Add a section about how the new exclude list may be a potential breaking
change.

klappradla commented:
Great point about the changelog @tessi 😹 

I added a small section - not too sure about the wording though...

tessi commented:
it's great. thanks!

pmeinhardt commented:
I am wondering whether we should give "Danger" a try on our open-source projects so we don't forget changelog entries and similar things…

![danger!](https://user-images.githubusercontent.com/706519/52403718-4a966080-2abf-11e9-94fc-d0c2c138a9bc.png)

— https://danger.systems/js/

klappradla commented:
It's probably a good idea in general @pmeinhardt 👍 

I just wonder: in theory there should be enough people looking at the PR to realize there's something missing 😉. I'm currently a bit skeptical about bot-tools on Github as "dependabot" is going on my nerves quite a bit on a different repo (and I don't have access to change its settings...).

While on this PR, I was actually wondering why we do not have CI for this project at all 🏜
